### PR TITLE
Webapp - Outgoing OAuth Connections

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR adds a new resource type in the integrations console, to be used for external OAuth systems. The feature is meant to support "Client Credentials" and "Password" grant types.

This pull request contains the following PRs:
- https://github.com/mattermost/mattermost/pull/25820
- https://github.com/mattermost/mattermost/pull/25822
- https://github.com/mattermost/mattermost/pull/25844
- https://github.com/mattermost/mattermost/pull/25899
- https://github.com/mattermost/mattermost/pull/25987

Enterprise PR https://github.com/mattermost/enterprise/pull/1626

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-55025
Fixes https://mattermost.atlassian.net/browse/MM-55030

#### Release Note

```release-note
Add Outgoing OAuth Connections integration type
```

### CODE REVIEW

**Comment by mickmister** [comment]
This PR should generally be treated as webapp-only, since I made no server changes on top of https://github.com/mattermost/mattermost/pull/25844

**Comment by fmartingr** [comment]
@mickmister Do we consider this the feature branch? If that's the case, did you modify any non-webapp code?

**Comment by fmartingr** [comment]
> This PR should generally be treated as webapp-only, since I made no server changes on top of #25844

Sorry I didn't read that. If that's the case, to make it easier for reviewers (code reviewers) you can point this branch to my last PR, so only that changes will show up. @esarafianou  made a good point about that in https://mattermost.atlassian.net/browse/MM-54990

**Comment by mickmister** [comment]
@fmartingr I've also updated this PR with master, so it's actually not in sync with your branch at the moment. Maybe we can do these merges to make them back in sync

`master -> MM-55034-validation-endpoint`

and

`MM-55034-validation-endpoint -> webapp-outgoing-oauth-connections`

**Comment by fmartingr** [comment]
> @fmartingr I've also updated this PR with master, so it's actually not in sync with your branch at the moment. Maybe we can do these merges to make them back in sync
> 
> `master -> MM-55034-validation-endpoint`
> 
> and
> 
> `MM-55034-validation-endpoint -> webapp-outgoing-oauth-connections`

I have updated all the train from master and conflicts are still showing up.

**Comment by mickmister** [comment]
@fmartingr Merged again, looks good now

**Comment by fmartingr** [comment]
> @fmartingr Merged again, looks good now

Ok I was trying to test this locally and I didn't see the Add button for the connections and the permissions were wrong, I noticed that the base branch is incorrect, the correct one is: `MM-56537-search-by-audience`. I merged that locally and the translations now show up but the button is still not there. Which steps should I follow to make it appear? (Permissions, FeatureFlag, Enterprise)

File: webapp/channels/src/components/integrations/outgoing_oauth_connections/installed_outgoing_oauth_connections.tsx
Comment (hmhealey) [inline webapp/channels/src/components/integrations/outgoing_oauth_connections/installed_outgoing_oauth_connections.tsx:58]: Can you also use `useIntl` in this component as well?

File: webapp/channels/src/packages/mattermost-redux/src/actions/integrations.ts
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/actions/integrations.ts:440]: Nit, this doesn't need `batchActions` since it's just a single action
```suggestion
        dispatch({
            type: IntegrationTypes.DELETED_OUTGOING_OAUTH_CONNECTION,
            data: {id},
        });
```

**Comment by mickmister** [comment]
@fmartingr Note that I had to make the changes in this commit to make the validation feature work https://github.com/mattermost/mattermost/pull/25507/commits/f371ba85baa70c983b6c891d41265c5b725e3f2b, as well as some changes in the enterprise repo [(branch comparison)](https://github.com/mattermost/enterprise/compare/MM-55033-outgoing-oauth-connection-logic..mattermost:MM-55033-outgoing-oauth-connection-logic_michaelk)

Summary of these changes:
- Monorepo:
    - Make the `/validate` endpoint accept a `id` in the payload, rather than a route parameter (we can make it a route param again though I thought this was simpler). If `id` is provided in payload, and `client_secret` is not provided in payload, then we use the stored connection's `ClientSecret` to do the validation
- Enterprise repo:
    - Add header `Content-Type: application/x-www-form-urlencoded` to token request
    - Handle case where OAuth provider returns an invalid response

**Comment by fmartingr** [comment]
> @fmartingr Note that I had to make the changes in this commit to make the validation feature work [f371ba8](https://github.com/mattermost/mattermost/commit/f371ba85baa70c983b6c891d41265c5b725e3f2b), as well as some changes in the enterprise repo [(branch comparison)](https://github.com/mattermost/enterprise/compare/MM-55033-outgoing-oauth-connection-logic..mattermost:MM-55033-outgoing-oauth-connection-logic_michaelk)
> 
> Summary of these changes:
> 
> * Monorepo:
>   
>   * Make the `/validate` endpoint accept a `id` in the payload, rather than a route parameter (we can make it a route param again though I thought this was simpler). If `id` is provided in payload, and `client_secret` is not provided in payload, then we use the stored connection's `ClientSecret` to do the validation
> * Enterprise repo:
>   
>   * Add header `Content-Type: application/x-www-form-urlencoded` to token request
>   * Handle case where OAuth provider returns an invalid response

Remember to create a PR for the enterprise changes with the same branch name as this PR, and run the `make i18n-extract` and all that stuff so enterprise CI passes.

**Review by hmhealey — APPROVED** [review]

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Nice work @mickmister. A few minor UI things I came across and some questions:

1. In the sidebar nav, it looks a bit odd when the text wraps on two lines for a nav item. This seems to be because of the line height. Can we change the following on sidebar items to fix this:
```
padding: 6px 12px 6px 34px;
line-height: 20px;
```
2. The breadcrumb at the top of the edit/add page has `Add` as the title of the parent page. It should say 'Outgoing OAuth 2.0 Connections`
    image
    image

3. When the user has their theme set to a dark theme, the title on the delete confirmation modal is getting a theme variable that doesn't work well in the backstage context. Also, I proposed some adjustments to this modal - let me know if they can be accommodated. And can the button be made red to indicate a destructive action?

**Implemented**

image


**Designed**

image

3. Is the `Grant type` selection no longer part of the requirement for outgoing oauth connections? 

4. Is the `Scopes` input no longer part of the requirement either?

5. Could you help me understand what the need for the `Validate` button is? I would expect we would do the validation upon hitting the `Save` button instead of needing another button. Seems unnecessary and unexpected to me. And it's also not great to have two primary buttons side-by-side like this.

6. I proposed UI improvements to the listing view. I assume we're not doing those since it affects all other list views in the integrations backstage? Or is it possible to include these improvements?

**Implemented**
image

**Designed**
image

**Comment by fmartingr** [comment]
Changed the base branch from `MM-55034-validation-endpoint` to `outgoing-oauth-connections-setting-toggle` to avoid showing unnecessary changes in between branches.

**Review by fmartingr — APPROVED** [review]
🥳

**Comment by mickmister** [comment]
@matthewbirtch Thanks for the thorough review :+1: Here are some thoughts on your points:

1. Done
2. Done
3. Done



4. Grant type "password" is not in scope of the first release of the feature. See https://hub.mattermost.com/private-core/pl/tk5z5k46n3g95ydy7g4rkathpr

5. Scopes were removed from the feature, as security said they were not relevant to this type of OAuth

6. For the validate feature, there are some tradeoffs depending on what we go with. First off, I want to show how Zoom does this to verify their webhook URL:

Zoom has a "Validate" button just below the URL input box, which sends a request to your webhook handler to ensure it's set up correctly. The user must click this button in order to validate the URL.

It used to be the case on the Zoom form that the "Save" button would be dimmed if you hadn't validated the URL first, though it seems they've changed it to instead show a warning after save, in the case that you hadn't validated the URL before save.

In Zoom's old implementation, I didn't like how the form wouldn't let me save before validating the URL. I now see value in preventing the save, since there is otherwise no forced confirmation that it is set up right. I'm 0/5 but leaning on both "separate validate button" and "save without validating" since the integration may not be set up yet. @hanzei Do you have any thoughts on this?

Validate button


Validate confirmation


Warning when validation step was not completed, after editing the webhook URL


One detail about our form, is that "saving" results in navigating to a different page/view, so it's not clear to me where the validation message should go, if the validation would be involved in the saving. Also, "validating" can potentially have side effects in the external system, so the developer probably wants to do it minimally. It basically just fetches a token to use, but then doesn't issue any requests with it afterwards. Another thing, the developer may just want to change the name of the OAuth connection later when editing the form, and in that case I don't think the implicit "validation" should be done when saving the form.

7. In the interest of keeping scope smaller I think we may not implement this. I will look into this. Some work for this was done here https://github.com/mattermost/mattermost-webapp/pull/11261

**Comment by mickmister** [comment]
@matthewbirtch The test cloud server we've been using is up-to-date with the PR

**Comment by mickmister** [comment]
> I'm still seeing this:

@matthewbirtch I thought I had applied your change and had the look of this fixed. I'll take another look :+1:

> It looks like the icons are not vertically aligned with the labels here either

I'm trying to resolve this but couldn't figure it out. Do you have any advice on how to align the icon with the message?

**Comment by matthewbirtch** [comment]
> @matthewbirtch Something else that comes to mind, when the validation fails, I would think that the `Validate Connection` should continue showing, so the user has an immediate way to revalidate. This is useful in the case that the user has entered all the values correctly, and the validation issue is on the integration side. At the moment, the user needs to edit one of the values related to the connection validation (client id, client secret, or token url) in order to show the validate button again.

That's a great point. I would suggest adding a `Try again` link to the error message like this

image

**Comment by mickmister** [comment]
@matthewbirtch The latest changes are running at https://outgoing-oauth-connections.test.mattermost.cloud if you want to take another look

**Comment by mickmister** [comment]
@matthewbirtch Do you have any advice on how documentation links should be shown in the UI? I'm thinking we should have a link for general configuration on the "Add/Edit Outgoing OAuth Connections" page, as well as on the "Add/Edit Slash Command" form, under the "Request URL" input box.

**Comment by matthewbirtch** [comment]
> @matthewbirtch Do you have any advice on how documentation links should be shown in the UI? I'm thinking we should have a link for general configuration on the "Add/Edit Outgoing OAuth Connections" page, as well as on the "Add/Edit Slash Command" form, under the "Request URL" input box.

- For the "Add/Edit Outgoing OAuth Connections" page, I think we can place help text with a link below the form
    image

- For the "Add/Edit Slash Command" form can we add a help text line under the existing?
    image

File: server/channels/api4/outgoing_oauth_connection.go
Comment (esarafianou) [inline server/channels/api4/outgoing_oauth_connection.go:356]: We need to use LimiterReader here, otherwise a really large payload could be sent (e.g. in the name) and cause performance/availability issues.  We can use the *c.App.Config().ServiceSettings.MaximumPayloadSizeBytes for the max size.

Please speak with @sbishel or @harshilsharma63 as they are currently working on something related and a unified approach is needed. Example: https://github.com/mattermost/mattermost/pull/25510/files

**Review by esarafianou — CHANGES_REQUESTED** [review]
Not done with the review, but sharing feedback as I encounter things

**Comment by esarafianou** [comment]
@mickmister another thing (non-security related) is that the UI is broken for a member with "manage outgoing oauth connections" permission but without "manage slash commands"
There's a Loading which should not exist as they don;t have access to view the exisiting oauth connections and the "add" button is missing. 

Screenshot 2024-02-07 at 13 59 42

**Comment by esarafianou** [comment]
Another QA issue with the UI of the validate feature

https://github.com/mattermost/mattermost/assets/17114347/31a5b546-3a56-4ac1-af76-a1e5a6f3c9d5

If someone clicks the Validate before adding the audiences, the validate fails but the validate button doesn't re-appear.

Speaking of which, the PR should go under QA before merging. I find QA issues by testing for security, but a proper QA testing is warranted.

File: server/channels/api4/outgoing_oauth_connection.go
Comment (fmartingr) [reply server/channels/api4/outgoing_oauth_connection.go]: Added `io.LimitedReader` to secure payload size here: https://github.com/mattermost/mattermost/pull/25507/commits/8ea552a82e99a14d40b8335c16e796706226686d

The provided solution was for more generic payloads, but the approach here is the same as they use in that PR.

**Comment by matthewbirtch** [comment]
> > Apart from having the Validate button re-appear, I think we need to move it below the audiences field, it's required for the validation.
> 
> I think this was bottom-right previously but it was moved to where it is now. cc: @matthewbirtch

I think we can still place it below the audience URL field like so:

image

And we can bring back the button when an error occurs

image

**Comment by DHaussermann** [comment]
@mickmister and @fmartingr I have done a round  of testing on this today. I then spoke with Michael about how the functional testing went and came up with these points :

- @DHaussermann to reproduce issues where validation hangs and were a 500 occurs with no error details on save
- As discussed with @mickmister, E2E test with error cases would be a huge benefit 
- Disable all autofill things in integrations backstage
- Issues with wildcard somepath/c*
- Improve client-side wildcard matching if needed

I will circle back to this tomorrow and try to repro the 2 issues from the first bullet (they did not occur when I tried to show Michael afterwards.)

**Comment by fmartingr** [comment]
> * @DHaussermann to reproduce issues where validation hangs and were a 500 occurs with no error details on save

Did you try to validate a new connection or an existing connection? Checked the code and the only code path that reaches a 500 is if the backend fails to retrieve the connection from the database or if the token fails to be retrieved (I probably need to update those http status codes).

> * Issues with wildcard somepath/c*

I have added better validation here: https://github.com/mattermost/enterprise/pull/1629/commits/dfca56142c9923923336799a64129d63a58746ea

File: server/channels/api4/outgoing_oauth_connection.go
Comment (esarafianou) [inline server/channels/api4/outgoing_oauth_connection.go:40]: Thinking more about the read permissions:

The `PermissionManageOutgoingOAuthConnections` makes sense. 

But why would we otherwise require **both** `PermissionManageOutgoingWebhooks` **and** `PermissionManageSlashCommands`. We could have a case where e.g. Team admins are granted slash commands but not outgoing webhooks permissions so that they can create slash commands. Slash commands don't require outgoing webhooks. Why would just the `PermissionManageSlashCommands` not be sufficient from granting them read access to the outgoing connections?

Comment (fmartingr) [reply server/channels/api4/outgoing_oauth_connection.go:40]: You are right. For read, you should only need one of the two, and even then, should we trim the output of the API? Since the permissions are intended only for usage, not for retrieving data.

Comment (esarafianou) [reply server/channels/api4/outgoing_oauth_connection.go:40]: Yes that should be the ideal. Do we need anything more than the name for the slash commands and the outgoing webhooks?

Comment (esarafianou) [reply server/channels/api4/outgoing_oauth_connection.go:40]: So with `PermissionManageOutgoingOAuthConnections` we don't need to trim, but with either `PermissionManageSlashCommands` or `PermissionManageOutgoingWebhooks`, we should respond only with the data needed

**Comment by mickmister** [comment]
> Apart from having the Validate button re-appear, I think we need to move it below the audiences field, it's required for the validation.

@esarafianou It's technically not required for the validation though. The form says that's required when you click validate because it checks if the form is filled out as it's gathering the fields for validation. If we put the validation button below the audience URL, it will look like you are validating the audience URL as well, but that's not what's happening here. I think the current UX is fine with the validate button

**Comment by mickmister** [comment]
@fmartingr CI is passing now other than these tests, and similar ones. Are you able to look into this? Also I think we should merge https://github.com/mattermost/enterprise/pull/1629

https://github.com/mattermost/mattermost/actions/runs/7842508597/job/21401068398?pr=25507

image

**Comment by esarafianou** [comment]
> @esarafianou It's technically not required for the validation though. The form says that's required when you click validate because it checks if the form is filled out as it's gathering the fields for validation. If we put the validation button below the audience URL, it will look like you are validating the audience URL as well, but that's not what's happening here. I think the current UX is fine with the validate button

As this is not security related, I'll leave that to you and Design. It was more a feedback as being an end user myself that I see the button after a few fields and before the audience one and I click it only to get an error that I need to fill an field after the button. But 0/5 on the final design.

**Comment by DHaussermann** [comment]
@mickmister I would agree to move the button down if you can easily adjust the scope of what it's doing. This confused me as well. The Validate button was not functional because audience was still blank.
image

**Comment by DHaussermann** [comment]
cc @fmartingr see the comment above ☝️  I did not test again after your latest commit but it does not sound related.

**Comment by fmartingr** [comment]
> cc @fmartingr see the comment above ☝️ I did not test again after your latest commit but it does not sound related.

Yeah I didn't change anything related to that. Would it make sense to leave it where it is but ignore the validaton for the audiences field? cc: @matthewbirtch

**Comment by matthewbirtch** [comment]
> > cc @fmartingr see the comment above ☝️ I did not test again after your latest commit but it does not sound related.
> 
> Yeah I didn't change anything related to that. Would it make sense to leave it where it is but ignore the validaton for the audiences field? cc: @matthewbirtch

@fmartingr, actually I think that makes more sense. I'm not sure the audience field needs to be a part of the validation for the oauth connection. Any objections @mickmister?

**Comment by fmartingr** [comment]
> > > cc @fmartingr see the comment above ☝️ I did not test again after your latest commit but it does not sound related.
> > 
> > 
> > Yeah I didn't change anything related to that. Would it make sense to leave it where it is but ignore the validaton for the audiences field? cc: @matthewbirtch
> 
> @fmartingr, actually I think that makes more sense. I'm not sure the audience field needs to be a part of the validation for the oauth connection. Any objections @mickmister?

In order to validate only client id, secret and URL are required. I'll defer to @mickmister decission here, but I think it makes sense and it will be less confusing for the users.

**Review by esarafianou — COMMENTED** [review]
Just two questions/comments for now. It should be good otherwise. Once answered, I'll approve.

File: server/channels/api4/outgoing_oauth_connection.go
Comment (fmartingr) [reply server/channels/api4/outgoing_oauth_connection.go:396]: In order to test that the connection is valid (and we can retrieve a token) for an already stored connection, the entire connection payload is sent from the frontend but the client secret (since the API does not expose that field), so we need to take the configuration from the frontend and append the client secret to validate the connection.

**Comment by mickmister** [comment]
@DHaussermann I changed the validate button logic to not require the audience field to be filled out, and kept the button to be under the token URL input

File: server/channels/api4/outgoing_oauth_connection.go
Comment (esarafianou) [reply server/channels/api4/outgoing_oauth_connection.go:396]: Great, thank you for the explanation.

**Review by esarafianou — APPROVED** [review]

**Review by DHaussermann — APPROVED** [review]
Tested and passed
I have re-tested after the latest commit.  
- Re-tested wildcarding functionality (as a precaution)
- Retested UI validation. I can see the latest change where validate is now specific to token URL as implied by the button placement
- Re-tested the creating of related slash commands and the new debouncing for the validation

No other known issues remain.
LGTM!  from a QA perspective. note https://github.com/mattermost/enterprise/pull/1626 is needed as well.

**Comment by mickmister** [comment]
@cwarnermm Merged developer docs is here 🙂  https://github.com/mattermost/mattermost-developer-documentation/pull/1334

I don't see this page on the dev docs site though https://developers.mattermost.com/integrate/slash-commands

Here's the preview site from the PR, which is merged:

image

Here is the actual site:

image

**Pull request merged**
